src/components/KTableData/KTableData.vue

```diff
@@ -172,91 +170,90 @@ import type {
   PageChangeData,
   PageSizeChangeData,
   TableState,
-  TableStatePayload,
   SwrvStateData,
   TableDataProps,
-  RowExpandPayload,
-  RowActionsTogglePayload,
+  TableColumnKey,
+  TableColumnVisibility,
+  TableColumnWidths,
+  TableDataEmits,
+  TableDataSlots,
 } from '@/types'
 import { EmptyStateIconVariants } from '@/types'
 import { getInitialPageSize, DEFAULT_PAGE_SIZE } from '@/utilities'
 
-const props = withDefaults(defineProps<TableDataProps>(), {
-  resizeColumns: false,
-  tablePreferences: () => ({}),
-  rowHover: true,
-  rowAttrs: () => ({}),
-  rowLink: () => ({} as RowLink),
-  rowBulkActionEnabled: () => true,
-  rowKey: '',
-  cellAttrs: () => ({}),
-  loading: false,
-  emptyStateTitle: 'No Data',
-  emptyStateMessage: 'There is no data to display.',
-  emptyStateActionMessage: '',
-  emptyStateIconVariant: EmptyStateIconVariants.Default,
-  error: false,
-  errorStateTitle: 'An error occurred',
-  errorStateMessage: 'Data cannot be displayed due to an error.',
-  errorStateActionMessage: '',
-  maxHeight: 'none',
-  hidePagination: false,
-  paginationAttributes: () => ({}),
-  rowExpandable: () => false,
-  rowExpanded: () => false,
-  hideHeaders: false,
-  nested: false,
-  hidePaginationWhenOptional: false,
-  hideToolbar: false,
-  /**
-   * KTableData props defaults
-   */
-  headers: () => ([]),
-  fetcherCacheKey: '',
-  cacheIdentifier: '',
-  searchInput: '',
-  initialFetcherParams: () => ({}),
-  clientSort: false,
-  sortable: true,
-})
+type ColumnKey = TableColumnKey<Header>
+type ColumnVisibility = TableColumnVisibility<Header>
+type ColumnWidths = TableColumnWidths<Header>
 
-const slots = useSlots()
+const {
+  resizeColumns,
+  tablePreferences = {},
+  rowHover = true,
+  rowAttrs = () => ({}),
+  rowLink = () => ({} as RowLink),
+  rowBulkActionEnabled = () => true,
+  rowKey: rowKeyProp,
+  cellAttrs = () => ({}),
+  loading,
+  emptyStateTitle = 'No Data',
+  emptyStateMessage = 'There is no data to display.',
+  emptyStateActionMessage = '',
+  emptyStateIconVariant = EmptyStateIconVariants.Default,
+  emptyStateButtonAppearance,
+  emptyStateActionRoute,
+  error,
+  errorStateTitle = 'An error occurred',
+  errorStateMessage = 'Data cannot be displayed due to an error.',
+  errorStateActionMessage = '',
+  errorStateActionRoute,
+  maxHeight = 'none',
+  hidePagination,
+  paginationAttributes = {},
+  rowExpandable = () => false,
+  rowExpanded = () => false,
+  hideHeaders,
+  nested,
+  hidePaginationWhenOptional,
+  hideToolbar,
+  headers = [],
+  fetcherCacheKey = '',
+  cacheIdentifier = '',
+  searchInput = '',
+  initialFetcherParams = {},
+  clientSort,
+  sortable = true,
+  fetcher,
+  sortHandlerFunction,
+  tooltipTarget,
+  ...restProps
```

Comment:
issue(?): This seems a bit opaque - how to know what other props are being applied here and in the `v-bind`? 

It may be a bit confusing to know what is being swallowed up here; if other props were added later in the TS file down the line they would automatically be applied via the `v-bind` in the template if the developer isn't careful to explicitly extract it.

Reply:
These are event handlers for `@cell:<event>` / `@row:<event>` events. We need to declare them in props otherwise we cannot get the listeners as we already declared them in `defineEmits`. I'll add a comment here.

Reply:
But wouldn't future props need to be added/extracted carefully?

Reply:
Updated.

Reply:
I'm not sure I see what was changed here?